File: public/icons/generate_text_icons.py
#!/usr/bin/env python3
"""
Kaleidorium Text Icon Generator
Generates PWA icons with "Kaleidorium" text instead of just "K"
"""

import os
from PIL import Image, ImageDraw, ImageFont
import sys

# Icon sizes to generate
ICON_SIZES = [
    (72, 'icon-72x72.png'),
    (96, 'icon-96x96.png'),
    (128, 'icon-128x128.png'),
    (144, 'icon-144x144.png'),
    (152, 'icon-152x152.png'),
    (192, 'icon-192x192.png'),
    (384, 'icon-384x384.png'),
    (512, 'icon-512x512.png'),
    (192, '../android-chrome-192x192.png'),
    (512, '../android-chrome-512x512.png'),
    (180, '../apple-touch-icon.png'),
    (32, '../favicon-32x32.png'),
    (16, '../favicon-16x16.png')
]

def create_rounded_rectangle_mask(size, radius):
    """Create a rounded rectangle mask."""
    mask = Image.new('L', (size, size), 0)
    draw = ImageDraw.Draw(mask)
    draw.rounded_rectangle([0, 0, size, size], radius=radius, fill=255)
    return mask

def generate_icon(size, filename):
    """Generate a single icon with Kaleidorium text."""
    
    # Settings
    text = "Kaleidorium"
    bg_color = (0, 0, 0)  # Black
    text_color = (255, 255, 255)  # White
    corner_radius = int(size * 0.2)  # 20% of size
    
    # Create image
    img = Image.new('RGBA', (size, size), bg_color + (255,))
    draw = ImageDraw.Draw(img)
    
    # Calculate font size - adjust this multiplier as needed
    font_size = max(int(size * 0.08), 8)
    
    # Try to load a good font, fall back to default
    font = None
    font_paths = [
        '/System/Library/Fonts/Helvetica.ttc',  # macOS
        '/Windows/Fonts/arial.ttf',  # Windows
        '/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf',  # Linux
        '/System/Library/Fonts/Times.ttc',  # macOS fallback
    ]
    
    for font_path in font_paths:
        try:
            if os.path.exists(font_path):
                font = ImageFont.truetype(font_path, font_size)
                break
        except:
            continue
    
    if font is None:
        try:
            font = ImageFont.load_default()
        except:
            font = ImageFont.load_default()
    
    # Get text dimensions
    bbox = draw.textbbox((0, 0), text, font=font)
    text_width = bbox[2] - bbox[0]
    text_height = bbox[3] - bbox[1]
    
    # Calculate position to center text
    x = (size - text_width) // 2
    y = (size - text_height) // 2
    
    # Draw text with subtle shadow for better readability
    if size > 32:  # Only add shadow for larger icons
        shadow_offset = max(1, size // 64)
        draw.text((x + shadow_offset, y + shadow_offset), text, 
                 font=font, fill=(0, 0, 0, 128))  # Semi-transparent shadow
    
    # Draw main text
    draw.text((x, y), text, font=font, fill=text_color)
    
    # Apply rounded corners if needed
    if corner_radius > 0:
        mask = create_rounded_rectangle_mask(size, corner_radius)
        img.putalpha(mask)
    
    return img

def main():
    """Generate all icons."""
    print("🎨 Generating Kaleidorium text icons...\n")
    
    # Check if PIL is available
    try:
        from PIL import Image, ImageDraw, ImageFont
    except ImportError:
        print("❌ PIL (Pillow) is required. Install with: pip install Pillow")
        sys.exit(1)
    
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    for size, filename in ICON_SIZES:
        try:
            img = generate_icon(size, filename)
            
            # Determine full path
            filepath = os.path.join(script_dir, filename)
            
            # Ensure directory exists
            os.makedirs(os.path.dirname(filepath), exist_ok=True)
            
            # Save the icon
            img.save(filepath, 'PNG', optimize=True)
            print(f"✅ Generated: {filename} ({size}x{size})")
            
        except Exception as e:
            print(f"❌ Failed to generate {filename}: {str(e)}")
    
    print("\n🎉 All icons generated successfully!")
    print("\n📝 Next steps:")
    print("1. Check the generated icons in your file explorer")
    print("2. Test your PWA to see the new text-based icons")
    print("3. Clear browser cache and reinstall PWA if needed")

if __name__ == "__main__":
    main()

File: public/icons/generate_text_icons.js
const fs = require('fs');
const path = require('path');

// Try to use canvas - install with: npm install canvas
let Canvas;
try {
    Canvas = require('canvas');
} catch (e) {
    console.log('Canvas not available. Please install with: npm install canvas');
    console.log('Or use the HTML generator: generate_kaleidorium_text_icons.html');
    process.exit(1);
}

const { createCanvas } = Canvas;

const iconSizes = [
    { size: 72, name: 'icon-72x72.png' },
    { size: 96, name: 'icon-96x96.png' },
    { size: 128, name: 'icon-128x128.png' },
    { size: 144, name: 'icon-144x144.png' },
    { size: 152, name: 'icon-152x152.png' },
    { size: 192, name: 'icon-192x192.png' },
    { size: 384, name: 'icon-384x384.png' },
    { size: 512, name: 'icon-512x512.png' }
];

// Also generate the main PWA icons
const mainIcons = [
    { size: 192, name: '../android-chrome-192x192.png' },
    { size: 512, name: '../android-chrome-512x512.png' },
    { size: 180, name: '../apple-touch-icon.png' },
    { size: 32, name: '../favicon-32x32.png' },
    { size: 16, name: '../favicon-16x16.png' }
];

const allIcons = [...iconSizes, ...mainIcons];

function generateIcon(size, fileName) {
    const canvas = createCanvas(size, size);
    const ctx = canvas.getContext('2d');
    
    // Settings
    const text = 'Kaleidorium';
    const backgroundColor = '#000000';
    const textColor = '#ffffff';
    const cornerRadius = 0.2; // 20%
    
    // Calculate font size based on icon size - optimized for Playfair Display
    // For "Kaleidorium" (11 characters), we want it to fit nicely with serif font
    const baseFontSize = size * 0.075; // Adjusted for serif font characteristics
    
    // Create rounded rectangle background
    ctx.fillStyle = backgroundColor;
    const radius = cornerRadius * (size / 2);
    
    // Draw rounded rectangle
    ctx.beginPath();
    ctx.roundRect(0, 0, size, size, radius);
    ctx.fill();
    
    // Set text properties - using serif fonts to match logo design
    ctx.fillStyle = textColor;
    ctx.font = `bold ${baseFontSize}px 'Playfair Display', 'Times New Roman', serif`;
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    
    // Add subtle text shadow to match the animated loading style
    ctx.shadowColor = 'rgba(255, 255, 255, 0.1)';
    ctx.shadowBlur = size / 48;
    ctx.shadowOffsetX = 0;
    ctx.shadowOffsetY = size / 128;
    
    // Draw text
    ctx.fillText(text, size / 2, size / 2);
    
    return canvas;
}

// Add roundRect polyfill
if (!CanvasRenderingContext2D.prototype.roundRect) {
    CanvasRenderingContext2D.prototype.roundRect = function(x, y, width, height, radius) {
        this.beginPath();
        this.moveTo(x + radius, y);
        this.lineTo(x + width - radius, y);
        this.quadraticCurveTo(x + width, y, x + width, y + radius);
        this.lineTo(x + width, y + height - radius);
        this.quadraticCurveTo(x + width, y + height, x + width - radius, y + height);
        this.lineTo(x + radius, y + height);
        this.quadraticCurveTo(x, y + height, x, y + height - radius);
        this.lineTo(x, y + radius);
        this.quadraticCurveTo(x, y, x + radius, y);
        this.closePath();
    };
}

function generateAllIcons() {
    console.log('🎨 Generating Kaleidorium text icons...\n');
    
    allIcons.forEach(iconInfo => {
        try {
            const canvas = generateIcon(iconInfo.size, iconInfo.name);
            const buffer = canvas.toBuffer('image/png');
            
            const filePath = path.join(__dirname, iconInfo.name);
            const dir = path.dirname(filePath);
            
            // Ensure directory exists
            if (!fs.existsSync(dir)) {
                fs.mkdirSync(dir, { recursive: true });
            }
            
            fs.writeFileSync(filePath, buffer);
            console.log(`✅ Generated: ${iconInfo.name} (${iconInfo.size}x${iconInfo.size})`);
        } catch (error) {
            console.error(`❌ Failed to generate ${iconInfo.name}:`, error.message);
        }
    });
    
    console.log('\n🎉 All icons generated successfully!');
    console.log('\n📝 Next steps:');
    console.log('1. Check the generated icons in your file explorer');
    console.log('2. Test your PWA to see the new text-based icons');
    console.log('3. Clear browser cache and reinstall PWA if needed');
}

// Run the generator
generateAllIcons();
